### ISSUE TITLE
Feature request: check for exception string (like Catch's CHECK_THROWS_WITH)

### ISSUE THREAD

**Issue by nlohmann** [onqtam/doctest#97]
## Description

I would like to have a macro equivalent to Catch's `CHECK_THROWS_WITH`, see https://github.com/philsquared/Catch/blob/master/docs/assertions.md#exceptions.

Currently, I would need to provide a `try`/`catch` block on my own and call `CHECK` on the exception's `what()` function.

**Comment by onqtam** [comment]
Matchers are on the roadmap but I won't have the time to implement them properly in the next 2-3 months - just signed a 3 month contract to get some funds... Until then you could use this solution:

```c++
#include "doctest.h"

#include <stdexcept>
#include <string>
#include <cstdio>

#define DOCTEST_THROWS_WITH_IMPL(the_assert, expr, message)                                        \
    do {                                                                                           \
        std::string _doctest_error_string;                                                         \
        bool        _doctest_threw = false;                                                        \
        try {                                                                                      \
            expr;                                                                                  \
        } catch(std::exception & e) {                                                              \
            _doctest_threw = true;                                                                 \
            if(std::string(e.what()) != message) {                                                 \
                _doctest_error_string = #expr;                                                     \
                _doctest_error_string +=                                                           \
                        "\nThe expression threw an exception with a different message!\n";         \
                _doctest_error_string += "  Expected: \"";                                         \
                _doctest_error_string += message;                                                  \
                _doctest_error_string += "\"\n";                                                   \
                _doctest_error_string += "  Received: \"";                                         \
                _doctest_error_string += e.what();                                                 \
                _doctest_error_string += "\"";                                                     \
            }                                                                                      \
        } catch(...) {                                                                             \
            _doctest_threw        = true;                                                          \
            _doctest_error_string = #expr;                                                         \
            _doctest_error_string += "\nThe expression threw an exception which does not derive "  \
                                     "from \"std::exception\"!";                                   \
        }                                                                                          \
        if(!_doctest_threw) {                                                                      \
            _doctest_error_string = #expr;                                                         \
            _doctest_error_string += "\nThe expression didn't throw at all!";                      \
        }                                                                                          \
        if(!_doctest_error_string.empty())                                                         \
            DOCTEST_EXPAND_VA_ARGS(the_assert)(__FILE__, __LINE__, _doctest_error_string.c_str()); \
    } while((void)0, 0)

#define DOCTEST_WARN_THROWS_WITH(expr, message) DOCTEST_THROWS_WITH_IMPL(DOCTEST_ADD_MESSAGE_AT, expr, message)
#define DOCTEST_CHECK_THROWS_WITH(expr, message) DOCTEST_THROWS_WITH_IMPL(DOCTEST_ADD_FAIL_CHECK_AT, expr, message)
#define DOCTEST_REQUIRE_THROWS_WITH(expr, message) DOCTEST_THROWS_WITH_IMPL(DOCTEST_ADD_FAIL_AT, expr, message)

#define WARN_THROWS_WITH DOCTEST_WARN_THROWS_WITH
#define CHECK_THROWS_WITH DOCTEST_CHECK_THROWS_WITH
#define REQUIRE_THROWS_WITH DOCTEST_REQUIRE_THROWS_WITH

TEST_CASE("using a custom exception assert macro") {
    SUBCASE("runtime_error") {
        CHECK_THROWS_WITH(throw std::runtime_error("omg"), "omg 2!!!");
    }
    SUBCASE("should pass") {
        REQUIRE_THROWS_WITH(throw std::runtime_error("wtf"), "wtf");
    }
    SUBCASE("non-std::exception related") {
        CHECK_THROWS_WITH(throw 5, "wtf");
    }
    SUBCASE("doesn't throw") {
        CHECK_THROWS_WITH(printf("side-effects\n"), "wtf");
    }
}
```

The output from this will look like this:

```
[doctest] doctest version is "1.2.5"
[doctest] run with "--help" for options
===============================================================================
C:\doctest\scripts\playground\test.cpp(52)
TEST CASE:  using a custom exception assert macro
  runtime_error

C:\doctest\scripts\playground\test.cpp(54) ERROR!
  throw std::runtime_error("omg")
The expression threw an exception with a different message!
  Expected: "omg 2!!!"
  Received: "omg"

===============================================================================
C:\doctest\scripts\playground\test.cpp(52)
TEST CASE:  using a custom exception assert macro
  non-std::exception related

C:\doctest\scripts\playground\test.cpp(60) ERROR!
  throw 5
The expression threw an exception which does not derive from "std::exception"!

side-effects
===============================================================================
C:\doctest\scripts\playground\test.cpp(52)
TEST CASE:  using a custom exception assert macro
  doesn't throw

C:\doctest\scripts\playground\test.cpp(63) ERROR!
  printf("side-effects\n")
The expression didn't throw at all!

===============================================================================
[doctest] test cases:      1 |      0 passed |      1 failed |      0 skipped
[doctest] assertions:      3 |      0 passed |      3 failed |
[doctest] Status: FAILURE!
```

**Comment by onqtam** [comment]
Matchers are on the roadmap but I won't have the time to implement them properly in the next 2-3 months - just signed a 3 month contract to get some cash... Until then you could use this solution:

```c++
#include "doctest.h"

#include <stdexcept>

#define DOCTEST_EXT_THROWS_WITH_IMPL(the_assert, expr, message)                                    \
    do {                                                                                           \
        doctest::String _doctest_exception_string;                                                 \
        try {                                                                                      \
            expr;                                                                                  \
        } catch(std::exception & e) { _doctest_exception_string = e.what(); } catch(...) {         \
            _doctest_exception_string = "Unknown exception";                                       \
        }                                                                                          \
        DOCTEST_EXPAND_VA_ARGS(the_assert)(_doctest_exception_string == doctest::String(message)); \
    } while((void)0, 0)

#define CHECK_THROWS_WITH(expr, message) DOCTEST_EXT_THROWS_WITH_IMPL(DOCTEST_CHECK, expr, message)
#define REQUIRE_THROWS_WITH(expr, message) DOCTEST_EXT_THROWS_WITH_IMPL(DOCTEST_REQUIRE, expr, message)

TEST_CASE("using a custom exception assert macro") {
    CHECK_THROWS_WITH(throw std::runtime_error("omg"), "omg 2!!!");
    REQUIRE_THROWS_WITH(throw std::runtime_error("wtf"), "wtf");
    CHECK_THROWS_WITH(throw 5, "wtf");
    CHECK_THROWS_WITH(((void)0), "wtf");
}
```

the output message will contain the exception result and what was expected - its not perfect, but should do the trick

**Comment by nlohmann** [comment]
Any news on this?

**Comment by onqtam** [comment]
This is a high priority feature. I'll try to get version 2.1 out by the end of the year but cannot promise it - will notify here when done. Thanks for waiting.

**Comment by onqtam** [comment]
This is now in the master branch!
in a few hours I'll update the changelog and make a github release.

So currently you can pass only C strings to the ```<LEVEL>_THROWS_WITH``` macros - like this:

```c++
CHECK_THROWS_WITH(do_parsing(), "bad parse!");
```

In the future support for matchers will be added (the ability to use objects instead of just plain old C strings) but the interface of the macros will not change.

**Comment by onqtam** [comment]
@nlohmann aaand a bit unrelated to this issue, but I just released version 2.2 and now the [**DOCTEST_CONFIG_SUPER_FAST_ASSERTS**](https://github.com/onqtam/doctest/blob/master/doc/markdown/configuration.md#doctest_config_super_fast_asserts) config option also affects the [**normal**](https://github.com/onqtam/doctest/blob/master/doc/markdown/assertions.md#expression-decomposing-asserts) asserts and not just the [**binary**](https://github.com/onqtam/doctest/blob/master/doc/markdown/assertions.md#binary-and-unary-asserts)) - see the results in the [**benchmarks**](https://github.com/onqtam/doctest/blob/master/doc/markdown/benchmarks.md#cost-of-an-assertion-macro)
